Release 93.0.0 (#2063)
MetaMask/core: packages/base-controller/CHANGELOG.md

## [Unreleased]

## [4.0.0]
### Added
- Add `ControllerGetStateAction` and `ControllerStateChangeEvent` types ([#1890](https://github.com/MetaMask/core/pull/1890), [#2029](https://github.com/MetaMask/core/pull/2029))
- Add `NamespacedName` type ([#1890](https://github.com/MetaMask/core/pull/1890))
  - This is the narrowest supertype of all names defined within a given namespace.

> 

Reviewer: ```suggestion
- Add `NotNamespacedBy` type ([#2051](https://github.com/MetaMask/core/pull/2051))

```

Author: Ah yes — covered in [68ac7c4](https://github.com/MetaMask/core/pull/2063/commits/68ac7c46c9776723959eb731c2dcc72710479793)

Reviewer: The entry in https://github.com/MetaMask/core/pull/2063/commits/68ac7c46c9776723959eb731c2dcc72710479793 is inaccurate, since the type does not return a positive result for non-namespaced names. I think the suggestion I have here more accurately describes the type.

### Changed
- **BREAKING:** Narrow controller messenger `ActionConstraint['handler']` type to remove usage of `any` ([#1890](https://github.com/MetaMask/core/pull/1890))

>   - This type is now defined as the universal supertype of all action handlers, meaning any functions can be safely assigned to it, regardless of argument types, number of arguments, or return value type.

Reviewer: ```suggestion
  - This type is now defined as the universal supertype of all functions, meaning any functions can be safely assigned as an action handler, regardless of argument types, number of arguments, or return value type.
```

> - **BREAKING:** Narrow controller messenger `ActionConstraint['handler']` type to remove usage of `any` ([#1890](https://github.com/MetaMask/core/pull/1890))
>   - This type is now defined as the universal supertype of all action handlers, meaning any functions can be safely assigned to it, regardless of argument types, number of arguments, or return value type.

Reviewer: ```suggestion
- **BREAKING:** Alter controller messenger `ActionConstraint['handler']` type to remove usage of `any` ([#1890](https://github.com/MetaMask/core/pull/1890))
  - This type is now defined as the universal supertype of all functions, meaning any functions can be safely assigned as an action handler, regardless of argument types, number of arguments, or return value type.
```

Author: Okay. I wanted to make sure this wasn't a problem in case we were wrong, but I guess we've checked this multiple times in various ways, so it's probably safe.

- **BREAKING:** Alter controller messenger `ActionHandler` type so `Action` type parameter must satisfy (updated) `ActionConstraint` ([#1890](https://github.com/MetaMask/core/pull/1890))
- **BREAKING:** Alter controller messenger `ExtractActionParameters` utility type so `Action` type parameter must satisfy (updated) `ActionConstraint` ([#1890](https://github.com/MetaMask/core/pull/1890))
- **BREAKING:** Alter controller messenger `ExtractEventHandler` utility type so `Event` type parameter must satisfy `EventConstraint` ([#1890](https://github.com/MetaMask/core/pull/1890))
- **BREAKING:** Alter controller messenger `ExtractEventPayload` utility type so `Event` type parameter must satisfy `EventConstraint` and `Event['payload']` must be an array (to match behavior of `ExtractEventHandler`) ([#1890](https://github.com/MetaMask/core/pull/1890))

> - **BREAKING:** Alter `controller messenger SelectorFunction` type so its `Event` parameter must satisfy `EventConstraint`, and so it returns a function whose arguments satisfy an event payload via `ExtractEventPayload` ([#1890](https://github.com/MetaMask/core/pull/1890))

Reviewer: ```suggestion
- **BREAKING:** Alter controller messenger `SelectorFunction` type so that its array type generic parameter `Args` is replaced by `Event`, which must satisfy `EventConstraint`, and it returns a function whose arguments satisfy an event payload type defined by its generic parameter `Event` ([#1890](https://github.com/MetaMask/core/pull/1890))
```

> - The restricted controller messenger now allows calling all internal events and actions ([#2050](https://github.com/MetaMask/core/pull/2050))

Reviewer: ```suggestion
- **BREAKING:** `BaseControllerV2` is renamed as `BaseController`. 
- **BREAKING:** The old `BaseController` class is marked as deprecated with the `@deprecated` tsdoc tag and exported as `BaseControllerV1`.
- The restricted controller messenger now allows calling all internal events and actions ([#2050](https://github.com/MetaMask/core/pull/2050))
```

Reviewer: Looks good!

>   - Previously internal events and actions were only usable if they were listed as "allowed". They are still permitted to be listed as "allowed" now, but it is no longer necessary.

Reviewer: ```suggestion
- **BREAKING:** Internal events and actions are no longer permitted to be listed as "allowed" in the restricted controller messenger constructor or the controller messenger `getRestricted` method ([#2051](https://github.com/MetaMask/core/pull/2051))
```

Author: I've incorporated this into [c2f90a1](https://github.com/MetaMask/core/pull/2063/commits/c2f90a17d6f7d105c40406884f468e34945aaea8), what do you think?

> - Bump `@metamask/utils` to ^8.2.0 ([#1957](https://github.com/MetaMask/core/pull/1957))

Author: Ah, good catch. This change is also breaking. I've also added an entry for `NotNamespacedBy`: [68ac7c4](https://github.com/MetaMask/core/pull/2063/commits/68ac7c46c9776723959eb731c2dcc72710479793)

> - Bump `@metamask/utils` to ^8.2.0 ([#1957](https://github.com/MetaMask/core/pull/1957))

Reviewer: Loosk good!

## [Unreleased]

## [4.0.0]
### Added
- Add `ControllerGetStateAction` and `ControllerStateChangeEvent` types ([#1890](https://github.com/MetaMask/core/pull/1890), [#2029](https://github.com/MetaMask/core/pull/2029))
- Add `NamespacedName` type ([#1890](https://github.com/MetaMask/core/pull/1890))
  - This is the narrowest supertype of all names defined within a given namespace.

> - Add `NotNamespacedBy` type, which matches a non-namespaced action/event name ([#2051](https://github.com/MetaMask/core/pull/2051))

Reviewer: ```suggestion
- Add `NotNamespacedBy` type, which matches an action/event name if and only if it is not prefixed by a given namespace ([#2051](https://github.com/MetaMask/core/pull/2051))
```
This is inaccurate, since the type does not return a positive result for non-namespaced names.

Reviewer: That's not actually true, it returns a positive match for non-namespaced names. But that's more of a side effect. 

The entry should reflect the fact that the type is primarily intended to match namespaced action/event names that are prefixed by a *different* namespace than the one that is given.
